index.js

```diff
@@ -804,234 +740,414 @@
 
     return isEmpty(values) ?
       [Unknown] :
-      or(reduce(values, types, refine), loose ? [Inconsistent] : []);
+      or(values.reduce(refine, types), loose ? [Inconsistent] : []);
   };
 
-  //  rejectInconsistent :: [Type] -> [Type]
+  //  rejectInconsistent :: Array Type -> Array Type
   var rejectInconsistent = function(types) {
-    return filter(types, function(t) {
-      return t.type !== 'INCONSISTENT' && t.type !== 'UNKNOWN';
+    return types.filter(function(t) {
+      return t.type !== INCONSISTENT && t.type !== UNKNOWN;
     });
   };
 
-  //  determineActualTypesStrict :: ([Type], [Type], [Any]) -> [Type]
+  //  determineActualTypesStrict ::
+  //    (Array Type, Array Type, Array Any) -> Array Type
   var determineActualTypesStrict = function(env, types, values) {
     var types$ = _determineActualTypes(false, env, types, [], values);
     return rejectInconsistent(types$);
   };
 
-  //  determineActualTypesLoose :: ([Type], [Type], [Any]) -> [Type]
+  //  determineActualTypesLoose ::
+  //    (Array Type, Array Type, Array Any) -> Array Type
   var determineActualTypesLoose = function(env, types, values) {
     var types$ = _determineActualTypes(true, env, types, [], values);
     return rejectInconsistent(types$);
   };
 
-  //  valuesToPairs :: ([Type], [Any]) -> [Pair Any [Type]]
-  var valuesToPairs = function(env, values) {
-    return map(values, function(x) {
-      return [x, determineActualTypesLoose(env, env, [x])];
+  //  TypeVarInfo = StrMap { types :: Array Type
+  //                       , values :: StrMap (Array Any) }
+
+  //  updateTypeVarInfo :: ... -> TypeVarInfo
+  var updateTypeVarInfo = function(
+    env,            // :: Array Type
+    typeVarInfo,    // :: TypeVarInfo
+    typeVar,        // :: Type
+    index,          // :: Integer
+    propPath,       // :: PropPath
+    values          // :: Array Any
+  ) {
+    var $typeVarInfo = {};
+    for (var typeVarName in typeVarInfo) {
+      var info = typeVarInfo[typeVarName];
+      var $values = {};
+      for (var key in info.values) $values[key] = info.values[key].slice();
+      $typeVarInfo[typeVarName] = {types: info.types.slice(), values: $values};
+    }
+    if (!hasOwnProperty.call($typeVarInfo, typeVar.name)) {
```

Comment:
I changed my mind. I found myself reluctant to replace `hasOwnProperty.call(o, k)` with `k in o` in various places because I was unsure of whether `o` was created via `Object.create(null)`. Without a type system it would be all too easy for a `{}` to slip through, and expecting contributors to mentally trace the ancestry of an object in order to determine the appropriate membership check is asking too much.

Although using `Object.create(null)` is "correct", ignoring inherited properties when performing membership checks is less likely to result in bugs than "promising" never to create objects with inherited properties.

I do wonder whether we should rewrite sanctuary-def in PureScript one day. ;)